src/transformers/modeling_flax_utils.py

```diff
@@ -170,46 +172,57 @@ class FlaxPreTrainedModel(PushToHubMixin, FlaxGenerationMixin):
+    # seems to dislike it. To go around this we add the correct type annotations under TYPE_CHECKING, so
+    # static analyzers pick this up, and fall back to untyped class variables at runtime.
+    if typing.TYPE_CHECKING:
+        config_class: ClassVar[Type[PretrainedConfig]]
+        base_model_prefix: ClassVar[str]
+        main_input_name: ClassVar[str] = "input_ids"
+    else:
+        config_class = None
+        base_model_prefix = None
+        main_input_name = "input_ids"
     _auto_class = None
     _missing_keys = set()
-
-    def __init__(
-        self,
-        config: PretrainedConfig,
-        module: nn.Module,
-        input_shape: Tuple = (1, 1),
-        seed: int = 0,
-        dtype: jnp.dtype = jnp.float32,
-        _do_init: bool = True,
-    ):
-        if config is None:
+    # dataclass attributes
+    _config: PretrainedConfig
+    _module: nn.Module
+    input_shape: Tuple[int, ...] = (1, 1)
+    seed: int = 0
+    dtype: jnp.dtype = jnp.float32
+    _do_init: bool = True
+    # derived fields
+    key: jax.random.KeyArray = dataclasses.field(init=False)
+    generation_config: Optional[GenerationConfig] = dataclasses.field(init=False)
+    _is_initialized: bool = dataclasses.field(init=False)
+    _params_shape_tree: Any = dataclasses.field(init=False)
+    _required_params: Set[Any] = dataclasses.field(init=False)
+    _params: Optional[Dict] = dataclasses.field(init=False)
+
+    def __post_init__(self):
+        if self._config is None:
             raise ValueError("config cannot be None")
 
-        if module is None:
+        if self._module is None:
             raise ValueError("module cannot be None")
 
-        # Those are private to be exposed as typed property on derived classes.
-        self._config = config
-        self._module = module
-
         # Those are public as their type is generic to every derived classes.
-        self.key = PRNGKey(seed)
-        self.dtype = dtype
-        self.input_shape = input_shape
-        self.generation_config = GenerationConfig.from_model_config(config) if self.can_generate() else None
+        self.key = PRNGKey(self.seed)
+        
+        
+        self.generation_config = GenerationConfig.from_model_config(self._config) if self.can_generate() else None
 
         # To check if the model was intialized automatically.
-        self._is_initialized = _do_init
+        self._is_initialized = self._do_init
 
-        if _do_init:
+        if self._do_init:
             # randomly initialized parameters
-            random_params = self.init_weights(self.key, input_shape)
+            random_params = self.init_weights(self.key, self.input_shape)
             params_shape_tree = jax.eval_shape(lambda params: params, random_params)
         else:
-            init_fn = partial(self.init_weights, input_shape=input_shape)
+            init_fn = partial(self.init_weights, input_shape=self.input_shape)
```

Comment:
Nice! Thanks for iterating on this logic, think it's simpler to understand like this